### PR TITLE
ipq806x: various fixes and improvements

### PR DESCRIPTION
**Pull request by Ansuel**
This pr contains various fix for the ipq806x platform
This pr contains: 
- It was notice that there is a problem with L2 cache scaling. This fixes that. (It has been proved by various user that have reported increased performance and stability)
- Update the cpufreq driver to use opp table and use a more standard way. Detailed description in commits message
- Rework tsense drive to fix a wrong error output and mem leak

Signed-off-by: Ansuel Smith <ansuelsmth@gmail.com>

### CODE REVIEW

**Comment by Ansuel** [comment]
@ynezz sorry if i bother you again...  can you review this?

**Comment by pkgadd** [comment]
I've been testing this for the last ~20 hours successfully on my nbg6817/ ipq8065.

`Tested-by: Stefan Lippers-Hollmann <s.l-h@gmx.de>`

**Comment by Ansuel** [comment]
I use this for 4 days without a problem 
I also received a private email that reported me that this fix a stability problem with some app (Tvheadend 4.3) that cause freeze and reboot on both kernel 4.14 and 4.19

**Comment by neheb** [comment]
I'm wondering if this improves ethernet support. Last I used my R7800 I was having throughput and latency issues.

**Comment by hnyman** [comment]
Some ideas:
* The L2 scaling issue commit message might contain a reference to the upstream commit/discussion (codeaurora(?)), forum discussion(?), from where the idea of changing the cache speed via the lowest speed surfaced.

* The tsense driver changes (thermal statistics) commit messages might have an explanation that the changes concern the thermal statistics. "tsense" is rather cryptic.

**Comment by Ansuel** [comment]
i really can't find now the commit but if i remember correctly... we find that a fix was merged to upstream related to another arm cpu that had the exact same problem. (Problem transitioning from one frequency to another and the necessity to revert to idle frequency before jumping to the new one)

About t sense... updated...

**Review by hnyman — COMMENTED** [review]
Nitpicking, but it should be "kryo", not "kyro" in the commit title

**Comment by SqTER-PL** [comment]
> I'm wondering if this improves ethernet support. Last I used my R7800 I was having throughput and latency issues.

I use this compilation over 3 days very intensively. Works better like a Master on 4.14 kernel. I didn't notice throughput and latency issues. NAT and USB work faster. Zero errors, very stable.

**Comment by SqTER-PL** [comment]
> I'm wondering if this improves ethernet support. Last I used my R7800 I was having throughput and latency issues.

I use this compilation on R7800 over 2 days very intensively. Works better like a Master on 4.14 kernel. I didn't notice throughput and latency issues. NAT and USB work faster. Zero errors, very stable.
@Ansuel - good job 👍

**Comment by Ansuel** [comment]
@hnyman i finally find the patch i was talking about... Added an extensive explanation about the L2 fix.

**Comment by shelterx** [comment]
Works great here too.

**Comment by Ansuel** [comment]
News on this?

**Comment by SqTER-PL** [comment]
It works very well, which is good news.

File: target/linux/ipq806x/patches-4.19/0052-PM-OPP-Update-the-voltage-tolerance-when-adjusting-t.patch [hunk 1]
```diff
@@ -19,20 +19,20 @@ Signed-off-by: Georgi Djakov <georgi.djakov@linaro.org>
 +++ b/drivers/opp/core.c
```
Comment (ynezz) [inline target/linux/ipq806x/patches-4.19/0052-PM-OPP-Update-the-voltage-tolerance-when-adjusting-t.patch]: This commit subject and description needs more love.

File: target/linux/ipq806x/patches-4.19/0049-PM-OPP-Support-adjusting-OPP-voltages-at-runtime.patch [hunk 2]
```diff
@@ -1,7 +1,14 @@
-From c949f08cf20fe82971fbdb4015daa38210da492e Mon Sep 17 00:00:00 2001
```
Comment (ynezz) [inline target/linux/ipq806x/patches-4.19/0049-PM-OPP-Support-adjusting-OPP-voltages-at-runtime.patch]: Commit subject/description should be improved.

File: target/linux/ipq806x/files-4.19/arch/arm/boot/dts/qcom-ipq8065.dtsi [hunk 3]
```diff
@@ -84,3 +31,88 @@
 		};
 	};
 };
+
+&opp_table0 {
+	compatible = "operating-points-v2-qcom-cpu";
+	nvmem-cells = <&speedbin_efuse>;
+
+	// On ipq8065 1.2 ghz freq is not present
```
Comment (ynezz) [inline target/linux/ipq806x/files-4.19/arch/arm/boot/dts/qcom-ipq8065.dtsi]: I would remove the comment and added it to the commit description, if you insist on the comment, then please use /**/ format.

File: target/linux/ipq806x/files-4.19/arch/arm/boot/dts/qcom-ipq8065.dtsi [hunk 4]
```diff
@@ -84,3 +31,88 @@
...
+		opp-microvolt-speed0-pvs0-v0 = <975000>;
+		opp-microvolt-speed0-pvs1-v0 = <950000>;
+		opp-microvolt-speed0-pvs2-v0 = <925000>;
+		opp-microvolt-speed0-pvs3-v0 = <900000>;
+		opp-microvolt-speed0-pvs4-v0 = <875000>;
+		opp-microvolt-speed0-pvs5-v0 = <825000>;
+		opp-microvolt-speed0-pvs6-v0 = <775000>;
+		opp-supported-hw = <0x1>;
+		clock-latency-ns = <100000>;
+	};
```
Comment (ynezz) [inline target/linux/ipq806x/files-4.19/arch/arm/boot/dts/qcom-ipq8065.dtsi]: What about newline between nodes?

File: target/linux/ipq806x/files-4.19/arch/arm/boot/dts/qcom-ipq8065.dtsi [hunk 5]
```diff
@@ -84,3 +31,88 @@
...
+		opp-microvolt-speed0-pvs1-v0 = <1225000>;
+		opp-microvolt-speed0-pvs2-v0 = <1200000>;
+		opp-microvolt-speed0-pvs3-v0 = <1175000>;
+		opp-microvolt-speed0-pvs4-v0 = <1150000>;
+		opp-microvolt-speed0-pvs5-v0 = <1100000>;
+		opp-microvolt-speed0-pvs6-v0 = <1050000>;
+		opp-supported-hw = <0x1>;
+		clock-latency-ns = <100000>;
+	};
+
```
Comment (ynezz) [inline target/linux/ipq806x/files-4.19/arch/arm/boot/dts/qcom-ipq8065.dtsi]: This newline is not needed.

File: target/linux/ipq806x/patches-4.19/0063-4-ip806x-tsense-rework-driver.patch [hunk 6]
```diff
@@ -0,0 +1,107 @@
+--- a/drivers/thermal/qcom/tsens-ipq8064.c
```
Comment (ynezz) [inline target/linux/ipq806x/patches-4.19/0063-4-ip806x-tsense-rework-driver.patch]: Line wrap the commit description to < 72 chars.

File: target/linux/ipq806x/files-4.19/arch/arm/boot/dts/qcom-ipq8064.dtsi [hunk 7]
```diff
@@ -573,10 +573,10 @@
 			#size-cells = <1>;
```
Comment (ynezz) [inline target/linux/ipq806x/files-4.19/arch/arm/boot/dts/qcom-ipq8064.dtsi]: `11 sensor` -> `11 sensors`
`reg propriety` -> `reg property`

File: target/linux/ipq806x/files-4.19/arch/arm/boot/dts/qcom-ipq8064.dtsi [hunk 8]
```diff
@@ -573,10 +573,10 @@
...
 			status = "okay";
 			tsens_calib: calib@400 {
-				reg = <0x400 0x10>;
+				reg = <0x400 11>;
 			};
 			tsens_backup: backup@410 {
-				reg = <0x410 0x10>;
+				reg = <0x410 11>;
```
Comment (ynezz) [inline target/linux/ipq806x/files-4.19/arch/arm/boot/dts/qcom-ipq8064.dtsi]: What about using hex notation?

**Review by ynezz — CHANGES_REQUESTED** [review]

**Comment by ynezz** [comment]
> Some ideas:
> 
>     * The tsense driver changes (thermal statistics) commit messages might have an explanation that the changes concern the thermal statistics. "tsense" is rather cryptic.

+1

**Comment by ynezz** [comment]
> It works very well, which is good news.

@SqTER-PL Which device?

File: target/linux/ipq806x/files-4.19/arch/arm/boot/dts/qcom-ipq8064.dtsi (hunk 8 above)
Comment (Ansuel) [reply target/linux/ipq806x/files-4.19/arch/arm/boot/dts/qcom-ipq8064.dtsi]: Changed to hex but i still think 11 better explain it...

If we change to hex here should we change also speedbin to 0x4 ?

File: target/linux/ipq806x/files-4.19/arch/arm/boot/dts/qcom-ipq8065.dtsi (hunk 3 above)
Comment (Ansuel) [reply target/linux/ipq806x/files-4.19/arch/arm/boot/dts/qcom-ipq8065.dtsi]: Think is better to put a comment in the dtss

File: target/linux/ipq806x/patches-4.19/0049-PM-OPP-Support-adjusting-OPP-voltages-at-runtime.patch (hunk 2 above)
Comment (Ansuel) [reply target/linux/ipq806x/patches-4.19/0049-PM-OPP-Support-adjusting-OPP-voltages-at-runtime.patch]: Is it good now? I put the merged patch

File: target/linux/ipq806x/patches-4.19/0052-PM-OPP-Update-the-voltage-tolerance-when-adjusting-t.patch (hunk 1 above)
Comment (Ansuel) [reply target/linux/ipq806x/patches-4.19/0052-PM-OPP-Update-the-voltage-tolerance-when-adjusting-t.patch]: Good now?

File: target/linux/ipq806x/patches-4.19/0049-PM-OPP-Support-adjusting-OPP-voltages-at-runtime.patch (hunk 2 above)
Comment (ynezz) [reply target/linux/ipq806x/patches-4.19/0049-PM-OPP-Support-adjusting-OPP-voltages-at-runtime.patch]: I've meant this one image as it's not possible to comment on this...

File: target/linux/ipq806x/patches-4.19/0052-PM-OPP-Update-the-voltage-tolerance-when-adjusting-t.patch (hunk 1 above)
Comment (ynezz) [reply target/linux/ipq806x/patches-4.19/0052-PM-OPP-Update-the-voltage-tolerance-when-adjusting-t.patch]: I've meant this one image as it's not possible to comment on it...

Comment (Ansuel) [reply target/linux/ipq806x/patches-4.19/0052-PM-OPP-Update-the-voltage-tolerance-when-adjusting-t.patch]: Sorry for the stupid question... How should i change them ?

Comment (ynezz) [reply target/linux/ipq806x/patches-4.19/0052-PM-OPP-Update-the-voltage-tolerance-when-adjusting-t.patch]: Imagine, that you're reading the Git history somewhere in 2021 and see `ipq806x: update 0052`, would you be able to tell from this commit subject what is this commit about?

Comment (ynezz) [reply target/linux/ipq806x/patches-4.19/0052-PM-OPP-Update-the-voltage-tolerance-when-adjusting-t.patch]: And this patch is probably some upstream backport, so perhaps your changes should be upstreamed first as well?

File: target/linux/ipq806x/patches-4.19/0049-PM-OPP-Support-adjusting-OPP-voltages-at-runtime.patch (hunk 2 above)
Comment (ynezz) [reply target/linux/ipq806x/patches-4.19/0049-PM-OPP-Support-adjusting-OPP-voltages-at-runtime.patch]: Perhaps something like this?

`ipq806x: opp/core: update patch allowing adjusting of OPP voltages at runtime`

File: target/linux/ipq806x/patches-4.19/0052-PM-OPP-Update-the-voltage-tolerance-when-adjusting-t.patch (hunk 1 above)
Comment (Ansuel) [reply target/linux/ipq806x/patches-4.19/0052-PM-OPP-Update-the-voltage-tolerance-when-adjusting-t.patch]: Thx for the explanation. 
About proposing the patch upstream i can try... Anyway i have updated it for now...

**Comment by Ansuel** [comment]
@ynezz everything should be good except for the voltage tolerance patch

**Comment by ynezz** [comment]
Thanks! Pulled into my staging tree at https://git.openwrt.org/openwrt/staging/ynezz.git

**Pull request closed without merging**